x-pack/plugins/security/server/authentication/authentication_service.ts

```diff
@@ -139,7 +155,43 @@ export class AuthenticationService {
       return t.notHandled();
     });
 
-    this.logger.debug('Successfully registered core authentication handler.');
+    http.registerOnPreResponse((request, preResponse, toolkit) => {
+      if (preResponse.statusCode !== 401 || !canRedirectRequest(request)) {
+        return toolkit.next();
+      }
+
+      if (!this.authenticator) {
+        // Core doesn't allow returning error here.
+        this.logger.error('Authentication sub-system is not fully initialized yet.');
+        return toolkit.next();
+      }
+
+      // If users can eventually re-login we want to redirect them directly to the page they tried
+      // to access initially, but we only want to do that for non-API routes. API routes that support
+      // redirects are solely used to support various authentication flows that wouldn't make any
+      // sense after successful authentication through login page.
+      const originalURL = !API_ROUTES_SUPPORTING_REDIRECTS.includes(request.route.path)
+        ? this.authenticator.getRequestOriginalURL(request)
```

Comment:
This might be contrived, but I'll say it anyway:

Let's say I navigate to `/app/foo#bar` as an authenticated user, and for whatever reason, I get a `401` response on that specific path. This will route me back to the login page (if available). The login page will see that I have a session, and send me to my `next` location, which is `/app/foo#bar`. That endpoint throws the same `401`, which repeats the cycle.

Now, the odds of a specific endpoint throwing a `401` for an otherwise valid session are probably pretty slim, but if this is possible, then it'd result in an infinite redirect loop.

Do you think it'd make sense to send the user to the `logout` endpoint first (if the login page is available), which then redirects to the `login` endpoint?

Reply:
This definitely can happen, but I cannot think of any case where it'd be intentional and I'm a bit scared that we can hide quite bad bugs and misuse trying to improve UX.

This can happen when, for example, we can authenticate request, but endpoint uses some other credentials to talk to ES or other service and doesn't handle the response properly. Or when user has both valid session and relies on `Authorization` header (browser plugins, proxies, or the fact that browsers can remember credentials if you enter them once during this session via native login prompt). Or if we happen to have bug in our authentication pipeline.

It'd also mean that we'll be optimizing behavior for the cases that shouldn't happen 99% of the time. **But** if we put my main concern aside and if you feel that's still worth handling I can propose just clearing session instead of additional redirect - luckily we have access to the request here and it's trivial to get access to the `session` instance as well (with a warning in the logs). This way we won't introduce any overhead when it's not needed, but unfortunately would also ignore SLO flows if any.

What do you think?

Reply:
> I mentioned that we have access to the session, but didn't realize initially that it'd also allow us to quickly know if there is an active session (or cookie to be precise, via getSID, no requests needed), and do proposed redirect only for those with the session. So yeah, if we decide to go this route, let's selectively redirect to the /logout as you suggested.

I like this plan! If it adds too much complexity, then we can defer this work and open an issue to track how painful this ends up being in practice

Reply:
It shouldn't be difficult to add, but the integration/functional test can be tricky (unless anonymous endpoint throwing 401 via test fixture plugin does what we need), let's see.